### PR TITLE
chore(android): gradle 8

### PR DESCRIPTION
**Pull request by AbdullahFaqeir**
chore(android): let there be gradle 8
chore(android): fixed deprecations in kroll-apt project
chore(android): made kroll-apt incremental to enhance build time

**GitHub issue:** https://github.com/tidev/titanium-sdk/issues/13579

This PR is dedicated to my mentor Nafe Abboushi.

### CODE REVIEW

**Review by m1ga — CHANGES_REQUESTED** [review]
Are the changes to the JAVA files necessary for the update? The more stuff that is changed, the more we have to test :wink: 

Also I wouldn't update the min and targetSDK with this PR. If we still can use 21 we should do that too. Same with target 34. I have a PR for that here https://github.com/tidev/titanium-sdk/pull/13940/ as the broadcast receiver has to be adjusted.

So if that is not needed for the gradle PR please keep it as it is and do it in a separate PR (less changes, less testing).

I couldn't build a SDK locally. It's almost working but the last step `titaniumPublication(MavenPublication)` tries to publish it online instead of making a local repo? I've canceled it as it was uploading stuff :smile: Didn't want to publish something I don't need.

There aren't all cmake version available during build process. I have a different PR about that https://github.com/tidev/titanium-sdk/pull/13966 and I had to use `3.22.1` as that is available on all platforms.

**Review by AbdullahFaqeir — COMMENTED** [review]
All reviews applied.

**Comment by m1ga** [comment]
I have to do some testing but the first round was very good and I was able to build kitchensink an my app now :+1: 

Review will follow once I have more time to test

File: android/app/build.gradle [hunk 1]
```diff
@@ -9,47 +9,50 @@ apply plugin: 'com.android.application'
...
-	lintOptions {
-		checkReleaseBuilds false
-	}
 	sourceSets {
 		main {
 			assets.srcDirs = [
-				'src/main/assets',
-				"${projectDir}/../titanium/build/outputs/ti-assets"
+					'src/main/assets',
+					"${projectDir}/../titanium/build/outputs/ti-assets"
```
Comment (cb1kenobi) [inline android/app/build.gradle:28]: Are these indentations intentional?

```suggestion
				'src/main/assets',
				"${projectDir}/../titanium/build/outputs/ti-assets"
```

File: android/templates/build/ti.constants.gradle [hunk 2]
```diff
@@ -6,12 +6,13 @@
...
-	tiAndroidXAppCompatLibVersion = '1.6.1'
+	tiNdkVersion = '26.2.11394342'
+	tiAndroidXAppCompatLibVersion = '1.4.1'
 	tiAndroidXCoreLibVersion = '1.9.0'
 	tiAndroidXFragmentLibVersion = '1.5.7'
 	tiMaterialLibVersion = '1.6.1'
 	tiPlayServicesBaseLibVersion = '18.2.0'
 	tiManifestPlaceholders = [
-		tiActivityConfigChanges: 'density|fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode'
+			tiActivityConfigChanges: 'density|fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode'
```
Comment (cb1kenobi) [inline android/templates/build/ti.constants.gradle:16]: Intentional indention?

```suggestion
		tiActivityConfigChanges: 'density|fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode'
```

File: android/titanium/build.gradle [hunk 3]
```diff
@@ -57,9 +59,9 @@ android {
 				// Set up "kroll-apt" @Kroll annotation processor to generate C/C++ code bindings between V8/Java.
 				// Also have it produce a JSON file of all bindings to be used by module build system.
 				arguments = [
-					'kroll.outputJsonFilePath': "${projectDir}/../../dist/android/titanium.bindings.json".toString(),
-					'kroll.outputCppDirPath': "${projectDir}/../runtime/v8/generated".toString(),
-					'kroll.jsModuleName': 'titanium'
+						'kroll.outputJsonFilePath': "${projectDir}/../../dist/android/titanium.bindings.json".toString(),
+						'kroll.outputCppDirPath'  : "${projectDir}/../runtime/v8/generated".toString(),
+						'kroll.jsModuleName'      : 'titanium'
```
Comment (cb1kenobi) [inline android/titanium/build.gradle:64]: Intentional indention?

```suggestion
					'kroll.outputJsonFilePath': "${projectDir}/../../dist/android/titanium.bindings.json".toString(),
					'kroll.outputCppDirPath'  : "${projectDir}/../runtime/v8/generated".toString(),
					'kroll.jsModuleName'      : 'titanium'
```

File: android/titanium/build.gradle [hunk 4]
```diff
@@ -82,50 +87,45 @@ android {
 		main {
 			manifest.srcFile 'AndroidManifest.xml'
 			assets.srcDirs = [
-				'assets',
-				"${projectDir}/../modules/ui/assets"
+					'assets',
+					"${projectDir}/../modules/ui/assets"
```
Comment (cb1kenobi) [inline android/titanium/build.gradle:91]: Intentional indention?

```suggestion
				'assets',
				"${projectDir}/../modules/ui/assets"
```

File: android/titanium/build.gradle [hunk 5]
```diff
@@ -82,50 +87,45 @@ android {
...
+					"${projectDir}/../modules/utils/src/java",
+					"${projectDir}/../modules/xml/src/java",
+					"${projectDir}/../runtime/common/src/java",
+					"${projectDir}/../runtime/v8/src/java"
 			]
 			res.srcDirs = [
-				'res',
-				"${projectDir}/../modules/ui/res"
+					'res',
+					"${projectDir}/../modules/ui/res"
```
Comment (cb1kenobi) [inline android/titanium/build.gradle:117]: Intentional indention?

```suggestion
				'res',
				"${projectDir}/../modules/ui/res"
```

File: android/titanium/build.gradle [hunk 6]
```diff
@@ -139,8 +139,8 @@ task updateV8Library() {
 		def v8MakeFile = file(v8MakeFilePath)
 		v8MakeFile.getParentFile().mkdirs()
 		v8MakeFile.text = [
-			"set(LIBV8_VERSION \"${packageJson.v8.version}\")",
-			"set(LIBV8_MODE \"${packageJson.v8.mode}\")"
+				"set(LIBV8_VERSION \"${packageJson.v8.version}\")",
+				"set(LIBV8_MODE \"${packageJson.v8.mode}\")"
```
Comment (cb1kenobi) [inline android/titanium/build.gradle:143]: Intentional indention?

```suggestion
			"set(LIBV8_VERSION \"${packageJson.v8.version}\")",
			"set(LIBV8_MODE \"${packageJson.v8.mode}\")"
```

File: android/untar.gradle [hunk 7]
```diff
@@ -27,10 +27,10 @@ task untar() {
...
 	// Use "ant" to extract the given tarball.
 	ant.untar(
-		compression: project.properties.compression,
-		overwrite: project.properties.overwrite,
-		src: project.properties.src,
-		dest: project.properties.dest)
+			compression: project.properties.compression,
+			overwrite: project.properties.overwrite,
+			src: project.properties.src,
+			dest: project.properties.dest)
```
Comment (cb1kenobi) [inline android/untar.gradle:33]: Intentional indention?

```suggestion
		compression: project.properties.compression,
		overwrite: project.properties.overwrite,
		src: project.properties.src,
		dest: project.properties.dest)
```

**Review by cb1kenobi — COMMENTED** [review]
This is AWESOME! I don't wanna lose momentum on this PR.

I found a few indention nitpicks. I haven't tested the PR yet, but I spent some time researching some of the changes such as no longer need to call `.setPackageName()`. Good stuff!

What sort of testing do we need to do to get this over the finish line? I can test on Windows doing emulator builds as well as build-only device/dist builds. I don't have any working Android devices. Sigh.

**Comment by m1ga** [comment]
> What sort of testing do we need to do to get this over the finish line?

:+1:  @cb1kenobi I would love to split this up into smaller steps as this is a big PR:
https://github.com/tidev/titanium-sdk/issues/13579#issuecomment-2021150043
I have made older PRs: first one only updates the gradle files to the newer syntax, 2nd one updates cmake and checkstyle (looks much but it's only linting). Then we only have the gradle 8 parts in here and can focus on that as it's a bigger change.

File: android/app/build.gradle [hunk 8]
```diff
@@ -9,47 +9,50 @@ apply plugin: 'com.android.application'
...
 // Set up Android app project.
 android {
-	compileSdkVersion 33
+	ndkVersion project.ext.tiNdkVersion
+	//namespace 'com.titanium.test'
 	defaultConfig {
 		applicationId 'com.titanium.test'
-		minSdkVersion 21
+		compileSdk 33
+		minSdkVersion 26
```
Comment (hansemannn) [inline android/app/build.gradle:17]: Doesn't this bump the default min-sdk version to Android 8? We should remain stable with the current one

File: android/app/build.gradle [hunk 9]
```diff
@@ -9,47 +9,50 @@ apply plugin: 'com.android.application'
 
 // Set up Android app project.
 android {
-	compileSdkVersion 33
+	ndkVersion project.ext.tiNdkVersion
+	//namespace 'com.titanium.test'
```
Comment (hansemannn) [inline android/app/build.gradle:13]: Remove comment

File: android/build.gradle [hunk 10]
```diff
@@ -6,17 +6,17 @@
...
+	ext.kotlin_version = '1.9.23'
 
 	repositories {
 		google()
 		mavenCentral()
 	}
 	dependencies {
-		classpath 'com.android.tools.build:gradle:7.0.4'
-		classpath 'com.google.gms:google-services:4.3.15'
+		classpath 'com.android.tools.build:gradle:8.3.1'
```
[the remark this answers is not in the document]
Comment (hansemannn) [reply android/build.gradle:16]: Let's do 8.10.1

File: android/cli/commands/_buildModule.js [hunk 11]
```diff
@@ -612,12 +612,6 @@ AndroidModuleBuilder.prototype.generateModuleProject = async function generateMo
 		this.logger.error('Unable to load Android <manifest/> content from "timodule.xml" file.');
 		throw err;
 	}
-	let packageName = moduleId;
-	if (packageName.indexOf('.') < 0) {
-		packageName = `ti.${packageName}`;
-	}
-	mainManifest.setPackageName(packageName);
-	await mainManifest.writeToFilePath(path.join(moduleMainDir, 'AndroidManifest.xml'));
```
Comment (hansemannn) [inline android/cli/commands/_buildModule.js:620]: What happened to this?

File: android/gradle/wrapper/gradle-wrapper.properties [hunk 12]
```diff
@@ -1,6 +1,7 @@
+#Tue Mar 19 21:28:34 EET 2024
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.4-bin.zip
```
Comment (hansemannn) [inline android/gradle/wrapper/gradle-wrapper.properties:4]: Use latest one as well

File: android/templates/build/root.build.gradle [hunk 13]
```diff
@@ -1,14 +1,13 @@
...
 
 	repositories {
 		google()
 		mavenCentral()
 	}
 	dependencies {
-		classpath 'com.android.tools.build:gradle:7.0.4'
-		classpath 'com.google.gms:google-services:4.3.10'
+		classpath 'com.android.tools.build:gradle:8.3.1'
+		classpath 'com.google.gms:google-services:4.4.1'
```
Comment (hansemannn) [inline android/templates/build/root.build.gradle:10]: Use latest ones for kotlin, gradle and google-services

**Review by hansemannn — CHANGES_REQUESTED** [review]
A few questions and changes we should verify. And can you please rebase with latest master to resolve the merge conflicts?

**Comment by AbdullahFaqeir** [comment]
Was this put in the system in any way or not, I hope so, and if not I would love to get back and work on it

**Comment by AbdullahFaqeir** [comment]
> A few questions and changes we should verify. And can you please rebase with latest master to resolve the merge conflicts?

Of course

File: android/app/build.gradle [hunk 14]
```diff
@@ -9,7 +9,8 @@ apply plugin: 'com.android.application'
 
 // Set up Android app project.
 android {
-	compileSdkVersion 33
```
Comment (m1ga) [inline android/app/build.gradle:12]: this is still needed. It complains that it is missing. It can be 34

**Comment by m1ga** [comment]
Did a first quick test. The current PR won't build. These files need to be changed:

android/modules/ui/src/java/ti/modules/titanium/ui/PickerProxy.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiImageView.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIActivityIndicator.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButton.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButtonBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIOptionBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUITabbedBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIText.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiRecyclerViewHolder.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/searchbar/TiUISearchBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIAbstractTabGroup.java

 Some `R.` resources need to be changed. E.g.
```
R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox_ExposedDropdownMenu
```
will be
```
com.google.android.material.R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox_ExposedDropdownMenu
```

We can't change the import `org.appcelerator.titanium.R` because some lines like `R.style.Widget_Titanium_OutlinedButton_IconOnly` are in the titanium namespace.

Best is to follow the errors when building the SDK (~30 changes). I'm not sure how to add files to this PR :blush:  But it won't take long to change it. Just go through the error messages and change `R` with `com.google.android.material.R`.

After that: SDK builds fine and I tested it with two apps with existing modules and they work fine. Kitchensink works too. Hyperloop-examples breaks: 
```
[ERROR] [GRADLE] A problem occurred configuring project ':gradle-project'.
[ERROR] [GRADLE] > Could not create an instance of type com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
[ERROR] [GRADLE]    > Namespace not specified. Specify a namespace in the module's build file. See https://d.android.com/r/tools/upgrade-assistant/set-namespace for information about setting the namespace.
[ERROR] [GRADLE]      
[ERROR] [GRADLE]      If you've specified the package attribute in the source AndroidManifest.xml, you can use the AGP Upgrade Assistant to migrate to the namespace value in the build file. Refer to https://d.android.com/r/tools/upgrade-assistant/agp-upgrade-assistant for general information about using the AGP Upgrade Assistant.
[ERROR] [GRADLE] 
```

File: android/gradle.properties [hunk 15]
```diff
@@ -14,4 +14,6 @@ android.useAndroidX=true
 android.enableJetifier=false
 
 # remove if com.android.tools.build:gradle can be raised in /android/build.gradle
+
 android.suppressUnsupportedCompileSdk=33,34
+
```
Comment (m1ga) [inline android/gradle.properties:19]: This can be removed now

**Comment by hansemannn** [comment]
@m1ga PR's can be checked out most easily with the Github CLI and `gh pr checkout 14014` in this case. It clones the branch and checks it out. You can then push into the branch and the commits occur here.

**Comment by m1ga** [comment]
@hansemannn nice! That worked :+1: SDK will build now, modules & hyperloop currently fail with the error above. 

@AbdullahFaqeir would be nice if you can find out where the namespace is missing. Apps without hyperloop work fine but I didn't check all controls yet.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButton.java [hunk 16]
```diff
@@ -60,23 +59,23 @@ public TiUIButton(final TiViewProxy proxy)
 		styleId = TiConvert.toInt(proxy.getProperty(TiC.PROPERTY_STYLE), styleId);
 		switch (styleId) {
 			case UIModule.BUTTON_STYLE_OPTION_POSITIVE:
-				styleId = R.attr.buttonBarPositiveButtonStyle;
+				styleId = com.google.android.material.R.attr.buttonBarPositiveButtonStyle;
```
Comment (hansemannn) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButton.java:62]: To keep the diff small, you could also just `import com.google.android.material.R;` I think.

**Comment by AbdullahFaqeir** [comment]
@m1ga @hansemannn I fixed it to build for modules, couldn't build tho on my Silicon machine, but it the fix for sure went after that stage.

**Comment by m1ga** [comment]
awesome! SDK builds, apps build, modules build :+1: We are getting closer.

* fix for hyperloop apps: https://github.com/tidev/hyperloop.next/pull/392
* small workflow update in https://github.com/tidev/titanium-sdk/pull/14112 (modules build fine, but when you start a example app build in a module it won't forward the `--sdk 12.6.0` parameter)

**Comment by m1ga** [comment]
one thing we'll need to tell the module developers: minSdk 12.6.0 is required when you build modules with this SDK as it is building with a higher Java version. The other way (old modules in 12.6.0 apps) works fine.

**Comment by AbdullahFaqeir** [comment]
> one thing we'll need to tell the module developers: minSdk 12.6.0 is required when you build modules with this SDK as it is building with a higher Java version. The other way (old modules in 12.6.0 apps) works fine.

@m1ga while generating the manifest file?

**Comment by m1ga** [comment]
when building an app with a module that was build with this PR you'll see
´Unsupported class file major version 61` as the module was build with JAVA_17 and the app is trying to build with JAVA_11 (current 12.4.0). But since we have the `minSdk` value in the module manifest it's fine. The developer just have to update it when using 12.6.0 to build the module

**Comment by AbdullahFaqeir** [comment]
> when building an app with a module that was build with this PR you'll see ´Unsupported class file major version 61`as the module was build with JAVA_17 and the app is trying to build with JAVA_11 (current 12.4.0). But since we have the`minSdk` value in the module manifest it's fine. The developer just have to update it when using 12.6.0 to build the module

@m1ga Exactly!

**Comment by AbdullahFaqeir** [comment]
@m1ga  @hansemannn I believe the PR is ready to be merged?

**Review by m1ga — COMMENTED** [review]
My tests look good:
* building the SDK :heavy_check_mark: 
* building app with multiple existing modules with this SDK :heavy_check_mark: 
* building modules and the included example app :heavy_check_mark: 
* building hyperloop-examples :heavy_check_mark: 
* building kitchensink-v2 :heavy_check_mark:

File: android/templates/module/generated/build.gradle [hunk 17]
```diff
@@ -18,7 +18,7 @@ android {
 	compileSdkVersion <%- compileSdkVersion %>
 	ndkVersion '22.1.7171670'
```
[the remark this answers is not in the document]
Comment (hansemannn) [reply android/templates/module/generated/build.gradle:20]: @AbdullahFaqeir You resolved this but added no commit?

[the remark this answers is not in the document]
Comment (AbdullahFaqeir) [reply android/templates/module/generated/build.gradle:20]: Ooh, didn't see yours, I believe yes it should, and it allows me to build on my machine, let me see this

[the remark this answers is not in the document]
Comment (AbdullahFaqeir) [reply android/templates/module/generated/build.gradle]: Done

**Review by hansemannn — APPROVED** [review]
LGTM!

**Review by m1ga — APPROVED** [review]

**Comment by AbdullahFaqeir** [comment]
LET'S GOOOO 🚀🚀🚀

**Comment by cb1kenobi** [comment]
Excellent work!

**Comment by AhmedMSayed** [comment]
Halla walla, That's Awesome @AbdullahFaqeir you saved my day

**Pull request merged**